File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque
def solution(bridge_length, weight, truck_weights):
    cnt = 0
    truck = deque(truck_weights)
    bridge = deque([0] * bridge_length)
    current_weight = 0
    while truck or current_weight > 0:
        cnt+=1
        current_weight -= bridge.popleft()
        if truck and current_weight + truck[0] <= weight:
            current_truck = truck.popleft()
            bridge.append(current_truck)
            current_weight += current_truck
        else:
            bridge.append(0)
        
    return cnt

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let bridge = []
    let bridgeWeight=0
    let time = 0
    while(bridge.length>0 || truck_weights.length>0) {
        for(let truck of bridge) truck[1]++
        for(let truck of bridge){
            if(truck[1]>bridge_length) bridgeWeight-=bridge.shift()[0]
        }
        if(bridgeWeight+truck_weights[0] <= weight) {
            let truck = truck_weights.shift()
            bridge.push([truck,1])
            bridgeWeight+=truck
        }
        time++
    }
    return time
}